`XXH_STATIC_ASSERT()` doesn't work with VC++2019 as a C compiler

When we compile a code with Visual C++ 2019 as a *C compiler*, it doesn't interpret `XXH_STATIC_ASSERT(0)` properly.

If we use Visual C++ 2019 as a *C++ compiler*, `XXH_STATIC_ASSERT(0)` works correctly though.

```C
/* C code */
#define XXH_STATIC_ASSERT(c)  do { enum { XXH_sa = 1/(int)(!!(c)) }; } while (0)

int main(int argc, char** argv) {
    XXH_STATIC_ASSERT(0); /* VC++2019 as a C compiler reports no error */
}
```

- [Test at Compiler Explorer](https://godbolt.org/z/Tqc18YPza)

```C
/* test.c or test.cpp */
int f() {
    enum { x = 123 / 0 };
    return x;
}
```

When we compile above code as `test.cpp`, VC++2019 reports the following error.

```
test.cpp(2,17): error C2131: expression did not evaluate to a constant
test.cpp(2,17): message : failure was caused by an undefined arithmetic operation
```

But compile as `test.c`, it reports nothing.

At Compiler Explorer, it seems VC++ as a C compiler is only compiler which accepts above `f()` without error / warning.

I don't know there's a compiler switch to disable this behavior.

----
(edit)
- `/std:c11`, `/std:c17` don't change the behavior.
- [`_Static_assert()`](https://docs.microsoft.com/en-us/cpp/c-language/static-assert-c) works.  But [it requires recent version of MSVC and Windows SDK](https://docs.microsoft.com/en-us/cpp/c-language/static-assert-c?view=msvc-160#requirements).  MSVC2019 isn't shipped with `<stdalign.h>` etc.

--- a/xxhash.h
+++ b/xxhash.h
@@ -1501,7 +1501,7 @@ static void* XXH_memcpy(void* dest, const void* src, size_t size)
 #  elif defined(__cplusplus) && (__cplusplus >= 201103L)            /* C++11 */
 #    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { static_assert((c),m); } while(0)
 #  else
-#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { char XXH_sa[(c) ? 1 : -1]; (void) XXH_sa; } while(0)
+#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { void xxh_sa(int x[1 - 2 * ((c) ? 0 : 1)]); (void) xxh_sa; } while(0)
 #  endif
 #  define XXH_STATIC_ASSERT(c) XXH_STATIC_ASSERT_WITH_MESSAGE(c,#c)
 #endif

--- a/xxhash.h
+++ b/xxhash.h
@@ -1494,7 +1494,17 @@ static void* XXH_memcpy(void* dest, const void* src, size_t size)
 #endif
 
 /* note: use after variable declarations */
-#define XXH_STATIC_ASSERT(c)  do { enum { XXH_sa = 1/(int)(!!(c)) }; } while (0)
+#ifndef XXH_STATIC_ASSERT
+#  if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 201112L)    /* C11 */
+#    include <assert.h>
+#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) static_assert((c),m)
+#  elif defined(__cplusplus) && (__cplusplus >= 201103L)            /* C++11 */
+#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) static_assert((c),m)
+#  else
+#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { struct XXH_sa { char xxh_static_assert[(c) ? 1 : -1]; }; } while (0)
+#  endif
+#  define XXH_STATIC_ASSERT(c) XXH_STATIC_ASSERT_WITH_MESSAGE((c),"XXH_STATIC_ASSERT")
+#endif
 
 /*!
  * @internal

--- a/xxhash.h
+++ b/xxhash.h
@@ -1501,9 +1501,9 @@ static void* XXH_memcpy(void* dest, const void* src, size_t size)
 #  elif defined(__cplusplus) && (__cplusplus >= 201103L)            /* C++11 */
 #    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { static_assert((c),m); } while(0)
 #  else
-#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { void xxh_sa(int x[1 - 2 * ((c) ? 0 : 1)]); (void) xxh_sa; } while(0)
+#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { struct xxh_sa { char x[(c) ? 1 : -1]; }; } while(0)
 #  endif
-#  define XXH_STATIC_ASSERT(c) XXH_STATIC_ASSERT_WITH_MESSAGE(c,#c)
+#  define XXH_STATIC_ASSERT(c) XXH_STATIC_ASSERT_WITH_MESSAGE((c),#c)
 #endif
 
 /*!

--- a/xxhash.h
+++ b/xxhash.h
@@ -1497,13 +1497,13 @@ static void* XXH_memcpy(void* dest, const void* src, size_t size)
 #ifndef XXH_STATIC_ASSERT
 #  if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 201112L)    /* C11 */
 #    include <assert.h>
-#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) static_assert((c),m)
+#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { static_assert((c),m); } while(0)
 #  elif defined(__cplusplus) && (__cplusplus >= 201103L)            /* C++11 */
-#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) static_assert((c),m)
+#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { static_assert((c),m); } while(0)
 #  else
-#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { struct XXH_sa { char xxh_static_assert[(c) ? 1 : -1]; }; } while (0)
+#    define XXH_STATIC_ASSERT_WITH_MESSAGE(c,m) do { char XXH_sa[(c) ? 1 : -1]; (void) XXH_sa; } while(0)
 #  endif
-#  define XXH_STATIC_ASSERT(c) XXH_STATIC_ASSERT_WITH_MESSAGE((c),"XXH_STATIC_ASSERT")
+#  define XXH_STATIC_ASSERT(c) XXH_STATIC_ASSERT_WITH_MESSAGE(c,#c)
 #endif
 
 /*!

